### PR TITLE
Efficiency improvements to processes table on Windows

### LINKED ISSUE #2574: windows: Slow queries on process table
Thanks for getting the Windows version out the door at last! Just wanted to report that querying the process table is pretty slow in the current build compared to native queries via `Powershell` or `WMIC`:

``` shell
PS C:\programdata\osquery> measure-command { & .\osqueryi.exe 'select * from processes;' --json }


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 14
Milliseconds      : 544
Ticks             : 145445951
TotalDays         : 0.000168340221064815
TotalHours        : 0.00404016530555556
TotalMinutes      : 0.242409918333333
TotalSeconds      : 14.5445951
TotalMilliseconds : 14544.5951
```

Here is the WInXP naive old-hat way to get the process table:

``` shell
PS C:\programdata\osquery> measure-command { & wmic PROCESS }


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 3
Milliseconds      : 624
Ticks             : 36248542
TotalDays         : 4.19543310185185E-05
TotalHours        : 0.00100690394444444
TotalMinutes      : 0.0604142366666667
TotalSeconds      : 3.6248542
TotalMilliseconds : 3624.8542
```

The slightly better WMI queries are faster

``` shell
PS C:\programdata\osquery> measure-command { Get-WmiObject -Query 'Select * From Win32_Process' }


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 103
Ticks             : 1036293
TotalDays         : 1.19941319444444E-06
TotalHours        : 2.87859166666667E-05
TotalMinutes      : 0.001727155
TotalSeconds      : 0.1036293
TotalMilliseconds : 103.6293
```

While the new hotness CIM Queries are slightly quicker

``` shell
PS C:\programdata\osquery> measure-command { Get-CimInstance -Query 'SELECT * From CIM_Process' }


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 97
Ticks             : 973313
TotalDays         : 1.12651967592593E-06
TotalHours        : 2.70364722222222E-05
TotalMinutes      : 0.00162218833333333
TotalSeconds      : 0.0973313
TotalMilliseconds : 97.3313
```

### PR DESCRIPTION
**Pull request by yying**
Removed the need for a WMI query for _every_ captured process ID.

<pre>
PS > measure-command { & .\osqueryi.exe 'select * from processes;' --json }

Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 178
Ticks             : 1788286
TotalDays         : 2.06977546296296E-06
TotalHours        : 4.96746111111111E-05
TotalMinutes      : 0.00298047666666667
TotalSeconds      : 0.1788286
TotalMilliseconds : 178.8286
</pre>

Closes #2574

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

**Comment by PoppySeedPlehzr** [comment]
ok to test

**Comment by theopolis** [comment]
This is a temporary speed up.

For tables that report PIDs, when used as the lhs operand, processes is expected to only select given 1 input PID. In this form you will `select * from processes` for each PID.

As a 'fair' test try:
```
osquery> .timer on
osquery> select count(1) from processes join processes using (pid);
+----------+
| count(1) |
+----------+
| 262      |
+----------+
Run Time: real 0.072 user 0.048000 sys 0.024000
```

**Comment by yying** [comment]
I'm not sure if a `processes` table on Windows using WMI can match that performance...

<pre>
PS > measure-command { Get-WmiObject -Query 'Select * From Win32_Process Where ProcessId=1588' }


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 74
Ticks             : 741041
TotalDays         : 8.57686342592593E-07
TotalHours        : 2.05844722222222E-05
TotalMinutes      : 0.00123506833333333
TotalSeconds      : 0.0741041
TotalMilliseconds : 74.1041
</pre>

It takes ~74ms just for querying one process, which doesn't differ too dramatically against querying everything:
<pre>
PS > measure-command { Get-WmiObject -Query 'Select * From Win32_Process' }


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 88
Ticks             : 882802
TotalDays         : 1.02176157407407E-06
TotalHours        : 2.45222777777778E-05
TotalMinutes      : 0.00147133666666667
TotalSeconds      : 0.0882802
TotalMilliseconds : 88.2802
</pre>

Perhaps, as @PoppySeedPlehzr mentioned, these "hot path" tables should use native APIs for better performance.

**Comment by osqueryer** [comment]
:no_good: The commit 5f6aaf0cb6fb65e6750c754db1262f889c181991 (Job results: [417](https://jenkins.osquery.io/job/osqueryPullRequestAudit/417/)) failed the code audit.

**Comment by PoppySeedPlehzr** [comment]
@yying, mind rebasing? #2573 made some changes to how the processes table is laid out.

File: osquery/tables/system/windows/processes.cpp [hunk 1]
```diff
@@ -36,85 +37,83 @@ std::set<long> getProcList(const QueryContext& context) {
...
+  r["pgroup"] = "-1";
+  r["uid"] = "-1";
+  r["euid"] = "-1";
+  r["suid"] = "-1";
+  r["gid"] = "-1";
+  r["egid"] = "-1";
+  r["sgid"] = "-1";
+
+  r["wired_size"] = "0";
+  r["resident_size"] = "0"; // Populate with WorkingSetSize (VT_BSTR)
```
Comment (theopolis) [inline osquery/tables/system/windows/processes.cpp]: Comment still needed?

File: osquery/tables/system/windows/processes.cpp [hunk 2]
```diff
@@ -36,85 +37,83 @@ std::set<long> getProcList(const QueryContext& context) {
...
+  r["cmdline"] = SQL_TEXT(sPlaceHolder);
+  s = result.GetString("ExecutionState", sPlaceHolder);
+  r["state"] = SQL_TEXT(sPlaceHolder);
+  s = result.GetLong("ParentProcessId", lPlaceHolder);
+  r["parent"] = BIGINT(lPlaceHolder);
+  s = result.GetLong("Priority", lPlaceHolder);
+  r["nice"] = INTEGER(lPlaceHolder);
+  r["on_disk"] = osquery::pathExists(r["path"]).toString();
+
+  // TODO: some of these such as cwd, wired_size, phys_footprint
```
Comment (theopolis) [inline osquery/tables/system/windows/processes.cpp]: Is this still true?

File: osquery/tables/system/windows/processes.cpp [hunk 3]
```diff
@@ -36,85 +37,83 @@ std::set<long> getProcList(const QueryContext& context) {
...
-  }
+void genProcess(const WmiResultItem& result, QueryData& results_data) {
+  Row r;
+  Status s;
+  long lPlaceHolder;
+  std::string sPlaceHolder;
+
+  s = result.GetLong("ProcessId", lPlaceHolder);
+  r["pid"] = s.ok() ? BIGINT(lPlaceHolder) : BIGINT(-1);
+  s = result.GetString("Name", sPlaceHolder);
```
Comment (theopolis) [inline osquery/tables/system/windows/processes.cpp]: For the second parameter, just use r["name"]. Ditch the s = in all of these if you're not using them.

Comment (yying) [inline osquery/tables/system/windows/processes.cpp]: Should be fixed in rebased version

File: osquery/tables/system/windows/processes.cpp [hunk 4]
```diff
@@ -41,107 +42,111 @@ std::set<long> getProcList(const QueryContext& context) {
...
-  for (const auto& pid : pidlist) {
-    genProcess(pid, results);
+  std::string query = "SELECT * FROM Win32_Process";
+
+  auto pidlist = getSelectedPids(context);
+  if (pidlist.size() > 0) {
+    std::vector<std::string> constraints;
+    for (const auto& pid : pidlist) {
+      constraints.push_back("ProcessId=" +
+                            boost::lexical_cast<std::string>(pid));
```
Comment (theopolis) [inline osquery/tables/system/windows/processes.cpp]: Sorry I didn't see this before, just a nit/note mostly, std::to_string is much faster than boost's lexical cast.

File: osquery/tables/system/windows/processes.cpp (hunk 2 above)
Comment (yying) [inline osquery/tables/system/windows/processes.cpp]: Nope.

File: osquery/tables/system/windows/processes.cpp (hunk 1 above)
Comment (yying) [inline osquery/tables/system/windows/processes.cpp]: Nope

**Pull request merged**